ouroboros-network/src/Ouroboros/Network/Protocol/BlockFetch/Codec.hs

```diff
@@ -109,6 +111,21 @@ codecBlockFetch encodeBlock     decodeBlock
         | not (LBS.null trailing) -> fail "trailing bytes in CBOR-in-CBOR"
         | otherwise               -> return (block payload)
 
+codecBlockFetchSerialised
+  :: forall block m.
+     MonadST m
+  => (HeaderHash block -> CBOR.Encoding)
+  -> (forall s. CBOR.Decoder s (HeaderHash block))
+  -> Codec (BlockFetch (SerialisedBlock block)) CBOR.DeserialiseFailure m LBS.ByteString
+codecBlockFetchSerialised =
+    codecBlockFetch encodeBlock (error "TODO ?")
+  where
+    -- Avoid converting to a strict ByteString, as that requires copying O(n)
+    -- in case the lazy ByteString consists of more than one chunks.
+    encodeBlock :: SerialisedBlock block -> CBOR.Encoding
+    encodeBlock = foldMap CBOR.encodePreEncoded
+                . LBS.toChunks
+                . serialisedBlockBytes
```

Comment:
This isn't quite right, but fortunately the right thing will also let us solve the TODO here.

The problem the TODO here has is that we want to return the encoded bytes, but cborg doesn't have a way to do that directly. Fortunately, we thought of this before and we made the message encoding for blocks be a CBOR bytestring containing a block (and same for the headers in chain sync). This means that we can in fact get the serialised representation of the block without decoding (we only decode the length prefix).

But this also shows that the `encodeBlock` is using the wrong encoding, since it's not using the CBOR-in-CBOR format that the `codecBlockFetch` is using. See `encodeBlockWrapped` and `decodeBlockWrapped` above.

This also shows that we really need a round trip property test for this codec, both self, and cross with the normal `codecBlockFetch` (both directions).

Reply:
It was using CBOR-in-CBOR, as it was calling `codecBlockFetch`, but it was not encoding the embedded payload correctly. I fixed it and have added property tests for the codec(s).